### PR TITLE
Serializable failure with recursion for DoTransaction on this error, and transaction aborted error handled

### PR DESCRIPTION
**Pull request by ice-myles**
Serializable failure with recursion for DoTransaction on this error, and transaction aborted error handled

### CODE REVIEW

File: connectors/storage/v2/api.go [hunk 1]
```diff
@@ -38,6 +39,11 @@ func DoInTransaction(ctx context.Context, db *DB, fn func(conn QueryExecer) erro
 			return nil, backoff.Permanent(err) //nolint:wrapcheck // Not needed.
 		}
 	})
+	if errors.Is(err, ErrSerializationFailure) {
```
Comment (ice-ares) [inline connectors/storage/v2/api.go]: err!=nill &&

File: connectors/storage/v2/api.go [hunk 2]
```diff
@@ -172,7 +178,9 @@ func IsUnexpected(err error) bool {
 		!IsErr(err, ErrRelationNotFound) &&
 		!IsErr(err, ErrNotFound) &&
 		!IsErr(err, ErrCheckFailed) &&
-		!IsErr(err, ErrRelationInUse)
+		!IsErr(err, ErrRelationInUse) &&
+		!IsErr(err, ErrSerializationFailure) &&
+		!IsErr(err, ErrTxAborted)
```
Comment (ice-ares) [inline connectors/storage/v2/api.go]: u don't retry this?

File: connectors/storage/v2/api.go [hunk 3]
```diff
@@ -206,8 +214,12 @@ func parseDBError(err error) error { //nolint:funlen // .
 
 			return terror.New(ErrCheckFailed, map[string]any{"column": column})
 		}
-
-		return err
```
Comment (ice-ares) [inline connectors/storage/v2/api.go:210]: I think u still need this, but after 222

File: connectors/storage/v2/api.go (hunk 2 above)
Comment (ice-myles) [reply connectors/storage/v2/api.go]: The question is if we should.
Something happened unexpected inside the requests (Get, Select, Exec) inside the transaction. It started to retry before. Now we just abort such transaction. Do we need to retry tx again? Seems like yes.

Comment (ice-ares) [reply connectors/storage/v2/api.go]: if it will work after a few retries, then yes

File: connectors/storage/v2/api.go (hunk 1 above)
Comment (ice-ares) [reply connectors/storage/v2/api.go]: add a check to make sure err is not null before

File: connectors/storage/v2/api.go [hunk 4]
```diff
@@ -206,8 +214,12 @@ func parseDBError(err error) error { //nolint:funlen // .
...
+		if dbErr.SQLState() == "40001" {
+			return terror.New(ErrSerializationFailure, map[string]any{})
```
Comment (ice-dionysos) [inline connectors/storage/v2/api.go]: I would return that error as is:
```
return ErrSerializationFailure
```

File: connectors/storage/v2/api.go [hunk 5]
```diff
@@ -206,8 +214,12 @@ func parseDBError(err error) error { //nolint:funlen // .
...
+		}
+		if dbErr.SQLState() == "25P02" {
+			return terror.New(ErrTxAborted, map[string]any{})
```
Comment (ice-dionysos) [inline connectors/storage/v2/api.go]: Same here.

**Review by ice-dionysos — CHANGES_REQUESTED** [review]

**Review by ice-dionysos — APPROVED** [review]

**Pull request merged**